torchao/prototype/float8nocompile/test/test.py

```diff
@@ -0,0 +1,67 @@
+import pytest
+import torch
+import torch.nn as nn
+
+from torchao.float8.float8_linear_utils import convert_to_float8_training
+
+from torchao.prototype.float8nocompile.float8nocompile_linear_utils import (
+    convert_to_float8_nocompile_training,
+)
+
+from torchao.utils import TORCH_VERSION_AT_LEAST_2_5
+
+if not TORCH_VERSION_AT_LEAST_2_5:
+    raise AssertionError("torchao.float8 requires PyTorch version 2.5 or greater")
+
+
+class TestModel(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.layers = nn.Sequential(
+            nn.Linear(32, 64, bias=False),
+            nn.Linear(64, 32, bias=False),
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return self.layers(x)
+
+
+@pytest.fixture
+def model1():
+    torch.manual_seed(0)
+    return TestModel()
+
+
+@pytest.fixture
+def model2():
+    torch.manual_seed(0)
+    return TestModel()
+
+
+def test_model_weights_and_gradients(model1, model2):
+    assert torch.cuda.is_available()
+    device = torch.device("cuda")
+
+    model1 = model1.to(torch.bfloat16).to(device)
+    model2 = model2.to(torch.bfloat16).to(device)
+
+    # compare production float8 linear conversion with no-compile version
+    convert_to_float8_nocompile_training(model1)
+    convert_to_float8_training(model2)
+
+    input_data = torch.randn(16, 32, dtype=torch.bfloat16).to(device)
+    loss_fn = nn.MSELoss()
+
+    output1 = model1(input_data)
+    output2 = model2(input_data)
+
+    loss1 = loss_fn(output1, torch.zeros_like(output1))
+    loss2 = loss_fn(output2, torch.zeros_like(output2))
+
+    loss1.backward()
+    loss2.backward()
+
+    # compare the weights and gradients of both models
```

Comment:
the weights should match since they matched at model creation and there is no optimizer step, so no need to test that

here are three things to test:
* output1 vs output2
* weight1.grad vs weight2.grad
* input1.grad vs input2.grad

Reply:
Ok done. I had to play around with how I cloned the input tensor to ensure the copies had both `requires_grad==True` and `is_leaf==True` so the gradients would not be `None`. The tests are passing now.

Reply:
For my own learning, is the reason we care about input tensor gradients because certain techniques like saliency maps rely on using the input gradients? I didn't consider this at first and was only considering the case of updating model weights, so I am just curious.